CONTRIBUTING.md

```diff
@@ -237,9 +237,83 @@ Spotless will write files to
 different copies of the formatted files, so that you can see how they
 differ and infer what is the problem.
 
-The `paddedCell()` option is disabled for normal operation in order to
-detect any misbehaviour. You can enabled the option from the command line
-by running Gradle with `-Dspotless.paddedcell`.
+The `paddedCell()` option is disabled for normal operation so that any
+misbehaviour is detected, and not just suppressed. You can enabled the
+option from the command line by running Gradle with `-Dspotless.paddedcell`.
+
+### JavaDoc
+
+Good JavaDoc can help with navigating and understanding code. Elasticsearch
+has some guidelines around when to write JavaDoc, and when not to.
+
+#### The short version
+
+   1. Always add JavaDoc to new code.
+   2. Add JavaDoc to existing code if you can.
+   3. Document the "why", not the "how", unless that's important to the
+      "why.
+   3. Don't document anything trivial or obvious (e.g. getters and
+      setters). In other words, the JavaDoc should add some value.
+
+#### The long version
+
+   1. If you add a new Java package, please also add package-level
+      JavaDoc that explains what the package is for. This can just be a
+      reference to a more foundational / parent package if appropriate. An
+      example would be a package hierarchy for a new feature or plugin -
+      the package docs could explain the purpose of the feature, any
+      caveats, and possibly some examples of configuration and usage.
+   2. New classes and interfaces must have class-level JavaDoc that
+      describes their purpose. There are a lot of classes in the
+      Elasticsearch repository, and it's easier to navigate when you
+      can quickly find out what is the purpose of a class. This doesn't
+      apply to inner classes or interfaces, unless you expect them to be
+      explicitly used outside their parent class.
+   3. New public methods must have JavaDoc, because they form part of the
+      contract between the class and its consumers. Similarly, new abstract
+      methods must have JavaDoc because they are part of the contract
+      between a class and its subclasses. It's important that contributors
+      know why they need to implement a method, and the JavaDoc should make
+      this clear.
```

Comment:
It's not clear to me from this section if a method which overrides a public abstract method should have javadoc or not, presuming the abstract method already does.  My gut instinct is "not unless it's doing something weird", but I'm open to other opinions.